charmcraft/parts.py

```diff
@@ -0,0 +1,209 @@
+        Allow pip to install of binary wheels.
+
+    Extra files to be included in the charm payload must be listed under
+    the ``prime`` file filter.
+    """
+
+    properties_class = CharmPluginProperties
+
+    @classmethod
+    def get_build_snaps(cls) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {
+            "python3-dev",
+            "python3-venv",
+            "python3-pip",
+            "python3-setuptools",
+            "python3-wheel",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        venv_dir = self._part_info.part_install_dir / self._part_info.venv_dir
+        return {
+            "PATH": "{}/bin:${{PATH}}".format(str(venv_dir)),
+        }
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        venv_dir = self._part_info.part_install_dir / self._part_info.venv_dir
+        pip_install_cmd = f"pip install --target={venv_dir}"
+        options = cast(CharmPluginProperties, self._options)
+        commands = []
+
+        if not options.charm_allow_pip_binary:
+            pip_install_cmd += " --no-binary :all:"
+
+        if options.charm_python_packages:
+            python_packages = " ".join(
+                [shlex.quote(pkg) for pkg in options.charm_python_packages]
+            )
+            python_packages_cmd = f"{pip_install_cmd} {python_packages}"
+            commands.append(python_packages_cmd)
+
+        if options.charm_requirements:
+            requirements = " ".join(f"-r {r!r}" for r in options.charm_requirements)
+            requirements_cmd = f"{pip_install_cmd} {requirements}"
+            commands.append(requirements_cmd)
+
+        install_dir = self._part_info.part_install_dir
+        commands.append(
+            'cp --archive --link --no-dereference . "{}"'.format(install_dir)
+        )
+
+        return commands
+
+
+def setup_parts():
+    """Initialize craft-parts plugins."""
+    plugins.register({"charm": CharmPlugin})
+
+
+def validate_part(data: Dict[str, Any]) -> None:
+    """Validate the given part data against common and plugin models.
+
+    :param data: The part data to validate.
+    """
+    if not isinstance(data, dict):
+        raise TypeError("value must be a dictionary")
+
+    spec = data.copy()
+    plugin_name = spec.get("plugin", "")
+    if not plugin_name:
+        raise ValueError("'plugin' not defined")
+
+    plugin_class = plugins.get_plugin_class(plugin_name)
+
+    # validate plugin properties
+    plugin_class.properties_class.unmarshal(spec)
+
+    # validate common part properties
+    plugins.strip_plugin_properties(spec, plugin_name=plugin_name)
```

Comment:
I'd maybe make the copy just prior to this to be obvious what's going to nuke it, or comment why the copy is necessary above.  Or perhaps change strip_plugin_properties() to return a new dict instead. No action required, just a thought.